### PR TITLE
PWN-6256 - Moonpay API. sell_transactions

### PR DESCRIPTION
**Pull request by gslevinkov**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-6256

## Description of Work

Moonpay calls split to two servers: client-side and server-side
- add proxy API calls
- add client-side API calls

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/splash/SplashPresenter.kt [hunk 1]
```diff
@@ -2,22 +2,29 @@ package org.p2p.wallet.splash
...
 ) : BasePresenter<SplashContract.View>(), SplashContract.Presenter {
 
     override fun attach(view: SplashContract.View) {
         super.attach(view)
         launch {
             moonpaySellRepository.loadMoonpayFlags()
             delay(MINIMUM_SPLASH_SHOWING_TIME_MS)
+
+            val response = moonpaySellRepository.getUserSellTransactions(tokenKeyProvider.publicKey.toBase58Instance())
+            Timber.i(response.toString())
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/splash/SplashPresenter.kt:27]: remove when api is ready

File: app/src/main/res/values/environment_config.xml [hunk 2]
```diff
@@ -17,9 +17,9 @@
     <string name="torusNetwork">testnet</string>
     <string name="torusLogLevel">debug</string>
 
-    <string name="moonpayWalletDomain">moonpay.wallet.p2p.org</string>
+    <string name="moonpayServerSideProxyBaseUrl">https://moonpay.wallet.p2p.org</string>
```
Comment (gslevinkov) [inline app/src/main/res/values/environment_config.xml:20]: is used for WebView and server-side calls that need secret key

File: app/src/main/res/values/environment_config.xml [hunk 3]
```diff
@@ -17,9 +17,9 @@
...
     <string name="moonpayBaseDomain">buy-staging.moonpay.com</string>
-    <string name="moonpayBaseUrl">https://api.moonpay.com/</string>
+    <string name="moonpayClientSideBaseUrl">https://api.moonpay.com/</string>
```
Comment (gslevinkov) [inline app/src/main/res/values/environment_config.xml:22]: requests that need only api key

File: app/src/main/java/org/p2p/wallet/infrastructure/network/NetworkModule.kt [hunk 4]
```diff
@@ -48,6 +48,11 @@ object NetworkModule : InjectionModule {
     const val DEFAULT_CONNECT_TIMEOUT_SECONDS = 30L
     const val DEFAULT_READ_TIMEOUT_SECONDS = 30L
 
+    enum class MoonpayRetrofitQualifier {
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/NetworkModule.kt:51]: I suggest to move enum classes to separate file

Comment (davran1001) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/NetworkModule.kt:51]: Also maybe we will separate network module, for example move Moonpay to separate module

Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/infrastructure/network/NetworkModule.kt:51]: makes sense, but i would leave it here for the developer to know which qualifer to use when he wants to inject retrofit and stumbled upon NetworkModule

File: app/src/main/java/org/p2p/wallet/infrastructure/network/environment/NetworkServicesUrlProvider.kt [hunk 5]
```diff
@@ -86,4 +87,22 @@ class NetworkServicesUrlProvider(
         }
         Timber.i("Torus environment changed and saved: $newUrl;$newVerifier;$newSubVerifier")
     }
+
+    fun loadMoonpayEnvironment(): MoonpayEnvironment {
+        val serverSideBaseUrl = sharedPreferences.getString(
+            KEY_MOONPAY_SERVER_SIDE_BASE_URL,
+            context.getString(R.string.moonpayServerSideProxyBaseUrl)
+        )!!
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/infrastructure/network/environment/NetworkServicesUrlProvider.kt:95]: orEmpty()

File: app/src/main/java/org/p2p/wallet/moonpay/model/MoonpayWidgetUrlBuilder.kt [hunk 6]
```diff
@@ -10,29 +12,26 @@ private const val QUERY_LOCK_AMOUNT = "lockAmount"
...
-            .appendQueryParameter(QUERY_API_KEY, moonpayApiKey)
+        val builder = Uri.parse(networkServicesUrlProvider.loadMoonpayEnvironment().buyWidgetUrl)
+            .buildUpon()
+            .appendQueryParameter(QUERY_API_KEY, BuildConfig.moonpayKey)
             .appendQueryParameter(QUERY_CURRENCY_CODE, tokenSymbol)
             .appendQueryParameter(QUERY_BASE_CURRENCY_AMOUNT, amount)
             .appendQueryParameter(QUERY_BASE_CURRENCY_CODE, currencyCode)
             .appendQueryParameter(QUERY_LOCK_AMOUNT, "false")
             .appendQueryParameter(QUERY_WALLET_ADDRESS, walletAddress)
             .apply { if (paymentMethod != null) appendQueryParameter(QUERY_PAYMENT_METHOD, paymentMethod) }
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/moonpay/model/MoonpayWidgetUrlBuilder.kt:33]: appendQueryParameter(QUERY_PAYMENT_METHOD, paymentMethod).takeIf { paymentMethod != null }

File: app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt [hunk 7]
```diff
@@ -46,4 +52,19 @@ class MoonpaySellRemoteRepository(
 
         return ipFlags.isSellAllowed
     }
+
+    @Throws(MoonpaySellError::class)
+    override suspend fun getUserSellTransactions(
+        userAddress: Base58String
+    ): List<MoonpaySellTransaction> = withContext(dispatchers.io) {
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/moonpay/repository/sell/MoonpaySellRemoteRepository.kt:59]: no reason to use dispatchers here, cause we are already in suspend context from retrofit functio

File: app/src/main/java/org/p2p/wallet/moonpay/model/MoonpayWidgetUrlBuilder.kt (hunk 6 above)
Comment (gslevinkov) [reply app/src/main/java/org/p2p/wallet/moonpay/model/MoonpayWidgetUrlBuilder.kt:33]: takeIf returns null 
apply consumes null, but also appendQueryParameter invoked anyways

**Pull request merged**